CMake build error/bugs using open62541-1.5.0-rc2

## Description

CMake build error/bugs using open62541-1.5.0-rc2. My amalgamated file is provided.

[open62541.c](https://github.com/user-attachments/files/24453434/open62541.c)

## Background Information / Reproduction Steps

Used CMake options:

```bash
-DUA_ENABLE_AMALGAMATION=ON \
-DUA_ARCHITECTURE=freertos-lwip \
-DUA_LOGLEVEL=100 \
-DUA_MULTITHREADING=0 \
-DUA_ENABLE_SUBSCRIPTIONS=ON \
-DUA_ENABLE_SUBSCRIPTIONS_EVENTS=OFF \
-DUA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS=OFF \
-DUA_ENABLE_METHODCALLS=ON \
-DUA_ENABLE_PARSING=ON \
-DUA_ENABLE_NODEMANAGEMENT=ON \
-DUA_ENABLE_IMMUTABLE_NODES=OFF \
-DUA_ENABLE_COVERAGE=OFF \
-DUA_ENABLE_DISCOVERY=ON \
-DUA_ENABLE_DISCOVERY_MULTICAST=OFF \
-DUA_ENABLE_DISCOVERY_SEMAPHORE=OFF \
-DUA_ENABLE_ENCRYPTION:STRING=MBEDTLS \
-DUA_ENABLE_ENCRYPTION_TPM2=OFF \
-DUA_NAMESPACE_ZERO=REDUCED \
-DUA_ENABLE_DIAGNOSTICS=OFF \
-DUA_ENABLE_JSON_ENCODING=OFF \
-DUA_ENABLE_TYPEDESCRIPTION=ON \
-DUA_ENABLE_STATUSCODE_DESCRIPTIONS=REDUCED \
-DUA_ENABLE_PUBSUB=ON \
-DUA_ENABLE_PUBSUB_FILE_CONFIG=OFF \
-DUA_ENABLE_PUBSUB_INFORMATIONMODEL=OFF \
```
In CMake UA_ENABLE_DIAGNOSTICS is set as OFF. tmpSession is a variable that is declared inside an ifdef block of UA_ENABLE_DIAGNOSTICS, but still used outside the ifdef block.
```
UA_StatusCode
getBoundSession(UA_Server *server, const UA_SecureChannel *channel,
                const UA_NodeId *token, UA_Session **session) {
...
    /* Session exists on another SecureChannel */
#ifdef UA_ENABLE_DIAGNOSTICS
    UA_Session *tmpSession = getSessionByToken(server, token);
    if(tmpSession)
        tmpSession->diagnostics.unauthorizedRequestCount++;
#endif

    /* The StatusCode indicates if the Session doesn't exist or whether it does
     * not match to SecureChannel */
    return (tmpSession) ?
        UA_STATUSCODE_BADSECURECHANNELIDINVALID :
        UA_STATUSCODE_BADSESSIONIDINVALID;
}
```

ua_types_encoding_json.h is only included if UA_ENABLE_JSON_ENCODING enabled but in this CMake this option is set as OFF:
```
if(UA_ENABLE_JSON_ENCODING)
    if(NOT UA_ENABLE_TYPEDESCRIPTION)
        message(FATAL_ERROR "JSON encoding requires parsing and type descriptions to be enabled")
    endif()
    list(APPEND lib_headers ${PROJECT_SOURCE_DIR}/deps/cj5.h
                            ${PROJECT_SOURCE_DIR}/src/ua_types_encoding_json.h)
    list(APPEND lib_sources ${PROJECT_SOURCE_DIR}/deps/cj5.c
                            ${PROJECT_SOURCE_DIR}/src/ua_types_encoding_json.c
                            ${PROJECT_SOURCE_DIR}/src/ua_types_encoding_json_105.c)
    if(UA_ENABLE_PUBSUB)
        list(APPEND lib_sources ${PROJECT_SOURCE_DIR}/src/pubsub/ua_pubsub_networkmessage_json.c)
    endif()
endif()
```
Inside the header file types of CtxJson and ParseCtx defined. In the final amalgamation file these types are used as a member of another struct.
```
#ifdef UA_ENABLE_PUBSUB
...
typedef struct {
    CtxJson ctx;
    UA_NetworkMessage_EncodingOptions eo;
} PubSubEncodeJsonCtx;

typedef struct {
    ParseCtx ctx;
    UA_NetworkMessage_EncodingOptions eo;
} PubSubDecodeJsonCtx;
...
UA_StatusCode
UA_NetworkMessage_encodeJsonInternal(PubSubEncodeJsonCtx *ctx,
                                     const UA_NetworkMessage *src);
#endif
```
PubSubEncodeJsonCtx is also used as a member of function prototype UA_NetworkMessage_encodeJsonInternal.

If my CMake is faulty please let me know.

## Checklist

Please provide the following information:

 - [+] open62541 Version (release number or git tag): open62541-1.5.0-rc2
 - [ ] Other OPC UA SDKs used (client or server): 
 - [ ] Operating system:
 - [ ] Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [ ] Wireshark network dump attached
 - [ ] Self-contained code example attached
 - [ ] Critical issue

Same issue here. Compilation error relating to `tmpSession`: 


```
[ 18%] Building C object CMakeFiles/open62541-object.dir/src/server/ua_server_binary.c.o
/usr/bin/cc -DUA_DYNAMIC_LINKING_EXPORT -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/deps -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/deps/nodesetLoader/src -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/ope
n62541/deps/nodesetLoader/backends/open62541/src -I/usr/include/libxml2 -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/include -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/plugins/include -I/home/roadster/dev/oss/open62541
-ruby/tmp/x86_64-linux/open62541/3.4.7/deps/open62541/src_generated -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/deps/nodesetLoader/include -I/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/deps/nodesetLoader/backends/open625
41/include -std=c99 -pipe -Wall -Wextra -Wpedantic -Wno-unused-parameter -Wno-overlength-strings -Wc++-compat -Wno-jump-misses-init -Wformat -Wformat-security -Wformat-nonliteral -Wmissing-prototypes -Wstrict-prototypes -Wredundant-decls -Wuninitialized -Winit-self -W
cast-qual -Wstrict-overflow -Wnested-externs -Wmultichar -Wundef -fno-strict-aliasing -fexceptions -ffunction-sections -fdata-sections -pthread -fPIC -O0 -g -MD -MT CMakeFiles/open62541-object.dir/src/server/ua_server_binary.c.o -MF CMakeFiles/open62541-object.dir/src
/server/ua_server_binary.c.o.d -o CMakeFiles/open62541-object.dir/src/server/ua_server_binary.c.o -c /home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/src/server/ua_server_binary.c
/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/src/server/ua_server_binary.c: In function ‘getBoundSession’:
/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/src/server/ua_server_binary.c:426:13: error: ‘tmpSession’ undeclared (first use in this function); did you mean ‘UA_Session’?
  426 |     return (tmpSession) ?
      |             ^~~~~~~~~~
      |             UA_Session
/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/src/server/ua_server_binary.c:426:13: note: each undeclared identifier is reported only once for each function it appears in
/home/roadster/dev/oss/open62541-ruby/ext/open62541/deps/open62541/src/server/ua_server_binary.c:429:1: warning: control reaches end of non-void function [-Wreturn-type]
  429 | }
      | ^
```

--- a/src/pubsub/ua_pubsub_networkmessage.h
+++ b/src/pubsub/ua_pubsub_networkmessage.h
@@ -16,7 +16,9 @@
 #include <open62541/pubsub.h>
 
 #include "../ua_types_encoding_binary.h"
-#include "../ua_types_encoding_json.h"
+#ifdef UA_ENABLE_JSON_ENCODING
+# include "../ua_types_encoding_json.h"
+#endif
 
 #ifdef UA_ENABLE_PUBSUB
 
@@ -33,6 +35,7 @@ typedef struct {
     UA_NetworkMessage_EncodingOptions eo;
 } PubSubDecodeCtx;
 
+#ifdef UA_ENABLE_JSON_ENCODING
 typedef struct {
     CtxJson ctx;
     UA_NetworkMessage_EncodingOptions eo;
@@ -42,6 +45,7 @@ typedef struct {
     ParseCtx ctx;
     UA_NetworkMessage_EncodingOptions eo;
 } PubSubDecodeJsonCtx;
+#endif
 
 const UA_FieldMetaData *
 getFieldMetaData(const UA_DataSetMessage_EncodingMetaData *emd,
@@ -126,9 +130,11 @@ UA_NetworkMessage_signEncrypt(UA_NetworkMessage *nm,
                               UA_Byte *encryptStart,
                               UA_Byte *sigStart);
 
+#ifdef UA_ENABLE_JSON_ENCODING
 UA_StatusCode
 UA_NetworkMessage_encodeJsonInternal(PubSubEncodeJsonCtx *ctx,
                                      const UA_NetworkMessage *src);
+#endif
 
 _UA_END_DECLS
 

